### PR TITLE
Dafny SBFT: model execution & checkpoint protocol

### PR DESCRIPTION
**Pull request by HristoStaykov**
* **Problem Overview**  
  Checkpoint generation/collection implemented in honest replicas. This gives the opportunity for us to limit the Working Window size. All infinite maps we used in the Working Window in Replicas are now converted to finite maps.
* **Testing Done**  
  This change is related to the formal model and does not affect the executable code.

### CODE REVIEW

File: docs/sbft-formal-model/messages.dfy [hunk 1]
```diff
@@ -17,7 +17,7 @@ module Messages {
   import opened HostIdentifiers
   import Network
 
-  type SequenceID = nat
+  type SequenceID = k:nat | 0 < k witness 1
```
Comment (teoparvanov) [inline docs/sbft-formal-model/messages.dfy:20]: Does `witness 1` indicate that k=1 is a witness that instantiates this predicate?

File: docs/sbft-formal-model/replica.i.dfy [hunk 2]
```diff
@@ -35,35 +38,45 @@ module Replica {
...
-      && PrePreparesRcvdWF(prePreparesRcvd)
+      && committedClientOperations.Keys == getActiveSequenceIDs(c)
+      && preparesRcvd.Keys == getActiveSequenceIDs(c)
+      && commitsRcvd.Keys == getActiveSequenceIDs(c)
+      && prePreparesRcvd.Keys == getActiveSequenceIDs(c)
+      && (forall x | x in prePreparesRcvd && prePreparesRcvd[x].Some? :: prePreparesRcvd[x].value.payload.PrePrepare?)
       && (forall seqID | seqID in preparesRcvd :: PrepareProofSetWF(c, preparesRcvd[seqID]))
       && (forall seqID | seqID in commitsRcvd :: CommitProofSetWF(c, commitsRcvd[seqID]))
     }
+    function ShiftWorkingWindow<T>(c:Constants, m:map<SequenceID,T>, empty:T) : map<SequenceID,T>
```
Comment (teoparvanov) [inline docs/sbft-formal-model/replica.i.dfy:72]: or simply, `Shift()`, as this function operates on the `workingWindow` structure?

File: docs/sbft-formal-model/replica.i.dfy [hunk 3]
```diff
@@ -473,14 +525,52 @@ module Replica {
...
+    && v' == v.(checkpointMsgsRecvd := v.checkpointMsgsRecvd.(msgs := v.checkpointMsgsRecvd.msgs + {msg}))
+  }
+
+  predicate HasStableCheckpointForSeqID(c:Constants, v:Variables, seqID:SequenceID) {
+    && v.WF(c)
+    && var relevantCheckpointMsgs := set msg | && msg in v.checkpointMsgsRecvd.msgs 
+                                               && msg.payload.CheckpointMsg?
+                                               && msg.payload.seqIDReached == seqID;
+    && |relevantCheckpointMsgs| >= c.clusterConfig.AgreementQuorum()
+  }
```
Comment (teoparvanov) [inline docs/sbft-formal-model/replica.i.dfy:551]: Awesome :) State of the art definition!

File: docs/sbft-formal-model/replica.i.dfy [hunk 4]
```diff
@@ -473,14 +525,52 @@ module Replica {
...
+
+  predicate AdvanceWorkingWindow(c:Constants, v:Variables, v':Variables, msgOps:Network.MessageOps<Message>, seqID:SequenceID) {
+    && v.WF(c)
+    && msgOps.NoSendRecv()
+    && seqID > v.workingWindow.lastStableCheckpoint
```
Comment (teoparvanov) [inline docs/sbft-formal-model/replica.i.dfy:556]: Shouldn't the pre-condition be stricter, and say something like `seqID > ls + win`, like in the paper - where `win` is the size of the active window?

**Review by teoparvanov — COMMENTED** [review]
Great progress introducing the checkpoint protocol, thanks @HristoStaykov ! 
Do the existing invariant proofs succeed with the Checkpoint protocol enabled?

File: docs/sbft-formal-model/messages.dfy (hunk 1 above)
Comment (HristoStaykov) [reply docs/sbft-formal-model/messages.dfy:20]: Yes, exactly. This is a change that we made because in our system we start committing from Sequence ID No:1. Previously in the model the **SequenceID** was **nat**, starting from 0 to infinity. Both ways the invariants are preserved.

File: docs/sbft-formal-model/replica.i.dfy (hunk 2 above)
Comment (HristoStaykov) [reply docs/sbft-formal-model/replica.i.dfy:72]: Yes, I will change it, now when I look at it, it is a bit too wordy in the usage :)

File: docs/sbft-formal-model/replica.i.dfy (hunk 4 above)
Comment (teoparvanov) [reply docs/sbft-formal-model/replica.i.dfy:556]: Ohh I see, good point.

**Review by teoparvanov — APPROVED** [review]
Looks great, thanks @HristoStaykov !
